Oops! All prosthetics! Adds -6 quirk with all prosthetic limbs + Icon Tutorial

KirbyDaMaster opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request
This PR adds a new negative quirk. You start the round with all 4 limbs as prosthetics. This is only -8, which is 2x as many negative points as just having one, so its not quite economical, but would definitely leave you screwed after an EMP

I also created a readme tutorial for adding custom SVG icons, with steps and recommended software.
<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game
Having the option to become a quad-amputee for a couple extra negative quirk points isnt quite over the top, but gives a little more risk reward to the mix. It opens you up to some really wacky shenanigans with EMPs and the medbay will just love you (not).

Also having a tutorial for SVG icon creation was an absolute necessity.

<!-- Argue for the merits of your changes and how they benefit the game, especially if they are controversial and/or far reaching. If you can't actually explain WHY what you are doing will improve the game, then it probably isn't good for the game in the first place. -->

## Changelog

<!-- If your PR modifies aspects of the game that can be concretely observed by players or admins you should add a changelog. If your change does NOT meet this description, remove this section. Be sure to properly mark your PRs to prevent unnecessary GBP loss. You can read up on GBP and it's effects on PRs in the tgstation guides for contributors. Please note that maintainers freely reserve the right to remove and add tags should they deem it appropriate. You can attempt to finagle the system all you want, but it's best to shoot for clear communication right off the bat. -->

:cl:
add: -6 Quirk for all prosthetics named "Quadruple Amputee"
change: prosthetic arm changed from -4 to -3
add: SVG icon tutorial readme to tgui tgfont folder
change: Minor spelling issues in negative quirks tab
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

KirbyDaMaster committed:
Working Icon, including tutorial for implementing SVG icon.

KirbyDaMaster committed:
Update .gitignore

KirbyDaMaster committed:
Delete .gitignore

necromanceranne commented:
Let me have all basic cyber organs as well. I want to fucking PERISH if EMP'd.

Mothblocks commented on tgui/packages/tgfont/README_ICON_TUTORIAL.txt:
```diff
@@ -0,0 +1,26 @@
+3. In ..\tgstation\tgui\packages\tgfont\dist you will find 3 files that were generated, a .css, .eot and a .WOFF2 file. Copy and replace all 3 of these files to
+    the following directory -> ..\tgstation\tgui\packages\tgfont\static
+
+4. Then we run/double click the BUILD.bat file in ..\tgstation\
```
wait why are you doing any of this. you should be able to just build. if you cant that is a bug

KirbyDaMaster replied:
> wait why are you doing any of this. you should be able to just build. if you cant that is a bug

Ask Anturk, hes the one managing the system and I followed his steps.

Mothblocks replied:
@AnturK theres no way you need to copy and replace files for tgfont bro

Mothblocks commented:
the writing on that guide is fairly sloppy but im also not sure its necessary, i think you were just overcomplicating your workflow

Time-Green commented on code/datums/quirks/negative.dm:
```diff
@@ -455,6 +457,47 @@
+	prostheticRA.replace_limb(human_holder)
+	prostheticLL.replace_limb(human_holder)
+	prostheticRL.replace_limb(human_holder)
+	qdel(old_LA, old_RA, old_LL, old_RL)
```
This is not ideal. Could you just make a list with the new types and use a for loop to replace the limbs? It’s also not necessary to store and delete all the old limbs, replace_limb probably has a paramater to just delete the replaced limb

DaedalusGame replied:
https://github.com/tgstation/tgstation/blob/master/code/modules/surgery/bodyparts/dismemberment.dm#L316 no parameter to delete the replaced limb, hence the existing quirk doing the same thing

Time-Green replied:
That code could also be better, but I think it’s a bit cruel to suggest an associative static list. I think doing what you suggested is good for now

Time-Green commented on tgui/packages/tgfont/static/tgfont.css:
```diff
@@ -1,7 +1,7 @@
 @font-face {
     font-family: "tgfont";
-    src: url("./tgfont.woff2?45c3c7acc69dd413375d77898d24e41e") format("woff2"),
```
What are you doing with the fonts?

KirbyDaMaster replied:
Its the only way to implement the icon. It is the method outlined by Anturk.

stylemistake replied:
when new icons are added, rebuild of fonts and css is required, which changes the hash of the font file

Time-Green commented on tgui/packages/tgfont/README_ICON_TUTORIAL.txt:
```diff
@@ -0,0 +1,26 @@
+
+If you plan on making your own SVG, i suggest Inkscape, as it is free and pretty powerful for vector graphics.
+
+1. Get whatever SVG you plan on using and put it in the ..\tgstation\tgui\packages\tgfont\icons folder.
```
You should atomize your PR’s, this is not very relevant to adding the quirk

KirbyDaMaster replied:
> You should atomize your PR’s, this is not very relevant to adding the quirk

Absolutely is relevant because if anyone wants to make a quirk in the future this is the only way to add a custom icon to the quirk.

Time-Green commented on .gitignore:
```diff
@@ -1,209 +0,0 @@
-**/node_modules
-
-# Screenshot tests
-/artifacts
```
Uh please don’t delete the gitignore

stylemistake replied:
what the fuck

ShizCalev commented:
Atomize.

tralezab commented on code/datums/quirks/negative.dm:
```diff
@@ -417,9 +417,11 @@
 	value = -4
 	mob_trait = TRAIT_PROSOPAGNOSIA
-	medical_record_text = "Patient suffers from prosopagnosia and cannot recognize faces."
+	medical_record_text = "Patient suffers from Prosopagnosia and cannot recognize faces."
```
this is not correct, prosopagnosia shouldn't be capitalized here

tralezab commented on code/datums/quirks/negative.dm:
```diff
@@ -98,7 +98,7 @@
 /datum/quirk/item_quirk/brainproblems
 	name = "Brain Tumor"
-	desc = "You have a little friend in your brain that is slowly destroying it. Better bring some mannitol!"
+	desc = "You have a little friend in your brain that is slowly destroying it. Better bring some Mannitol!"
```
again, no capitalization needed

Iamgoofball replied:
but it's referring to a proper noun, Mannitol

Time-Green replied:
Medicines are not proper nouns, they’re just trivial names to easily refer to compounds. It’s not a person, place or thing

tralezab commented on tgui/packages/tgfont/README_ICON_TUTORIAL.txt:
```diff
@@ -0,0 +1,26 @@
+The following is the process to implement your own icon using an svg.
```
What everyone else has been saying, atomize this out. Means move it to a different pr

VaelophisNyx commented:
yog coder here to suggest the alternative: just make each limb selectable for -2 with no incompatibilities outside the random 1 limb quirk like we did over on yog?

KirbyDaMaster commented:
> Why add the icon tutorial in a completely unrelated pr?

It absolutely is required, you cannot add a custom icon for the quirk without following those exact steps. It was not described elsewhere.

tralezab commented:
you know what if kirby says it's related and required I'll keep it in

optimumtact commented:
who the hell are these atomizing andy's in here? the readme is directly relevant and is due to an undocumented process that is required.

adding it in this PR is fine?

KirbyDaMaster commented:
> the writing on that guide is fairly sloppy but im also not sure its necessary, i think you were just overcomplicating your workflow

Sloppy how? please elaborate.

KirbyDaMaster committed:
Optimizations, as well as blacklisting quirks from working with each other.

KirbyDaMaster committed:
Spelling adjust

removed capitalization of mannitol

KirbyDaMaster committed:
Spelling adjustment

removed capitalization of illness

DaedalusGame committed:
Revert "Delete .gitignore"

This reverts commit 5b640de8182b21f30788b56e6e31dcf49cbf7ba0.

DaedalusGame committed:
Revert "Update .gitignore"

This reverts commit 19e0fb9b80d2dcf93c1259a46c6ff5cebcd863f6.

DaedalusGame committed:
- Move the text changes to the other PR

KirbyDaMaster committed:
removed #4 from the tutorial, 2nd build was not needed.

KirbyDaMaster committed:
Merge branch 'Prosthetic-Quirk' of https://github.com/KirbyDaMaster/tgstation into Prosthetic-Quirk

Mothblocks commented on code/datums/quirks/negative.dm:
```diff
@@ -447,14 +449,37 @@
+	for(var/i in 1 to 4)
+		var/path = prosthetics[i]
+		var/obj/item/bodypart/prosthetic = new path
+		var/obj/item/bodypart/old_limbs = human_holder.get_bodypart(zones[i])
```
loop over zones directly instead

Mothblocks commented on code/datums/quirks/negative.dm:
```diff
@@ -447,14 +449,37 @@
+		qdel(old_limbs)
+
+/datum/quirk/quadruple_amputee/post_add()
+	to_chat(quirk_holder, "<span class='boldannounce'>All your limbs have been replaced with surplus prosthetics. They are fragile and will easily come apart under duress. Additionally, \
```
use `span_boldannounce()`

KirbyDaMaster replied:
made the change

Mickyan commented on code/datums/quirks/negative.dm:
```diff
@@ -447,14 +449,37 @@
+	desc = "Oops! All Prosthetics! Due to some truly cruel cosmic punishment, all your limbs have been taken from you."
+	icon = "tg-prosthetic-full"
+	value = -8
+	medical_record_text = "During physical examination, patient was found to have all prosthetic limbs"
```
Missing period at the end

KirbyDaMaster replied:
changed

Mickyan commented:
Not too fond of this because prosthetic limbs can almost be considered sidegrades to regular limbs (no wounds, easy to repair trading for EMP weakness and lower max health) and the prosthetic limb quirk is one of those most people already consider free points

Personally I would lower the point value to -6 or so, but this definitely needs to be blacklisted from Frail at the very least

RuneQuester commented:
SO CLOSE to being able to choose WHICH limb is prosthetic. we are SO close

KirbyDaMaster committed:
Merge branch 'Prosthetic-Quirk' of https://github.com/KirbyDaMaster/tgstation into Prosthetic-Quirk

DaedalusGame commented on code/datums/quirks/negative.dm:
```diff
@@ -447,14 +449,37 @@
+/datum/quirk/quadruple_amputee/add_unique()
+
+	var/mob/living/carbon/human/human_holder = quirk_holder
+	var/list/zones = list(BODY_ZONE_L_ARM, BODY_ZONE_R_ARM, BODY_ZONE_L_LEG, BODY_ZONE_R_LEG)
+	var/list/prosthetics = list(/obj/item/bodypart/l_arm/robot/surplus, /obj/item/bodypart/r_arm/robot/surplus, /obj/item/bodypart/l_leg/robot/surplus, /obj/item/bodypart/r_leg/robot/surplus)
+	for(var/i in 1 to 4)
+		var/path = prosthetics[i]
+		var/obj/item/bodypart/prosthetic = new path
+		var/obj/item/bodypart/old_limbs = human_holder.get_bodypart(zones[i])
+		prosthetic.replace_limb(human_holder, special = TRUE)
+		qdel(old_limbs)
```
I don't see this at all chief

- replace_body_part isn't an existing proc, replace_limb is but doesn't qdel the old limb
- if it doesn't matter whether the old limb is qdel'd we should just use replace_limb instead
- otherwise you're asking for a new proc to be created that will be confusing to people getting started with limb code and will only be used in like 1 or 2 places

KirbyDaMaster replied:
updated

KirbyDaMaster committed:
Several changes

Added replace_body_part to _HELPERS
Changed <span> to be span_boldannounce()
Blacklisted frail quirk from quad amputee amd paraplegic

KirbyDaMaster commented:
> Not too fond of this because prosthetic limbs can almost be considered sidegrades to regular limbs (no wounds, easy to repair trading for EMP weakness and lower max health) and the prosthetic limb quirk is one of those most people already consider free points
> 
> Personally I would lower the point value to -6 or so, but this definitely needs to be blacklisted from Frail at the very least

resolved

KirbyDaMaster committed:
fixed build errors.

Kapu1178 commented on code/__HELPERS/bodyparts.dm:
```diff
@@ -1 +1,6 @@
 #define IS_ORGANIC_LIMB(limb) (limb.bodytype & BODYTYPE_ORGANIC)
+
+/proc/replace_body_part(mob/living/carbon/human/human_holder, zone, obj/item/bodypart/prosthetic)
+	var/obj/item/bodypart/old_limb = human_holder.get_bodypart(zone)
+	prosthetic.replace_limb(human_holder, special = TRUE)
+	qdel(old_limb)
```
```suggestion
/mob/living/carbon/proc/replace_body_part(mob/living/carbon/human/human_holder, zone, obj/item/bodypart/prosthetic)
	var/obj/item/bodypart/old_limb = human_holder.get_bodypart(zone)
	prosthetic.replace_limb(human_holder, special = TRUE)
	qdel(old_limb)
```

DaedalusGame replied:
cool but you forgot to change the holder from parameter to src

JohnFulpWillard replied:
``src.get_bodypart``
💀

JohnFulpWillard commented:
Please handle @Kapu1178 's reviews

KirbyDaMaster committed:
Update code/__HELPERS/bodyparts.dm

Co-authored-by: Kapu1178 <75460809+Kapu1178@users.noreply.github.com>

KirbyDaMaster committed:
KapuKapuKapuKapuKapuKapuKapuKapuKapu

KirbyDaMaster reviewed (commented):
I think what is here is fine.

DaedalusGame committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into Prosthetic-Quirk

# Conflicts:
#	code/datums/quirks/negative.dm

Mothblocks commented on code/__HELPERS/bodyparts.dm:
```diff
@@ -1 +1,7 @@
 #define IS_ORGANIC_LIMB(limb) (limb.bodytype & BODYTYPE_ORGANIC)
+
+/mob/living/carbon/proc/del_and_replace_bodypart(obj/item/bodypart/new_limb, special)
```
1. Document this
2. Put this in the mob folder, __HELPERS should be kept somewhat slim/global

DaedalusGame replied:
Document where

Mothblocks replied:
A triple slashed comment makes a doc comment

```
/// Doc comment goes here
/thing/proc/action()
```

code/modules/mob/living/carbon, and make like carbon_bodyparts.dm

DaedalusGame replied:
why not put it in dismember code

Mothblocks replied:
Sure just not in a helper

DaedalusGame replied:
Done

DaedalusGame committed:
- More shedding

Mothblocks commented on tgui/packages/tgfont/README_ICON_TUTORIAL.txt:
```diff
@@ -0,0 +1,24 @@
+The following is the process to implement your own icon using an svg.
+
+If you plan on making your own SVG, i suggest Inkscape, as it is free and pretty powerful for vector graphics.
+
+1. Get whatever SVG you plan on using and put it in the ..\tgstation\tgui\packages\tgfont\icons folder.
+
+2. In VScode, at the top, open "Terminal>New Terminal" and run this exact line of code without the quotes -> ".\BUILD.cmd tg-font"
+    Wait till its all complete
+
+3. In ..\tgstation\tgui\packages\tgfont\dist you will find 3 files that were generated, a .css, .eot and a .WOFF2 file. Copy and replace all 3 of these files to
+    the following directory -> ..\tgstation\tgui\packages\tgfont\static
+
+Now your SVG will be able to be used in the game.
+
+NOTE; Be sure when you reference your icon that you prefix it with "tg-", otherwise it will not find it.
+
+EXAMPLE;
+    for svg named "prosthetic-leg.svg", you would reference it like so;
+        vvvvvvvvvvvvvvvvvvvvvvvvvv
+        icon = "tg-prosthetic-leg"
+        ^^^^^^^^^^^^^^^^^^^^^^^^^^
+    The process uses the main filename of your SVG as the icon name, so don't name it something stupid.
+
+    Also try to get your SVG as simple as possible, the engine has trouble rendering SVGs that have a lot of little disconnected parts
```
```suggestion
If you plan on making your own SVG, consider [Inkscape](https://inkscape.org/). It is free and pretty powerful for vector graphics.

1. Get whatever SVG you plan on using and put it in the `tgstation\tgui\packages\tgfont\icons` folder.

2. In VS Code, press Ctrl+Shift+B, and select "tgui: rebuild tgfont". Wait for it to comlpete.

Now your SVG will be able to be used in the game.

When you reference your icon that you prefix it with "tg-", otherwise it will not find it. For example, with an SVG named "prosthetic-leg.svg", you would reference it with `icon_state = "tg-prosthetic-leg"`.

Keep your SVG as simple as possible, the engine has trouble rendering SVGs that have a lot of little disconnected parts.
```

CC @AnturK to make sure I have the new instructions right.

General writing touch ups.

KirbyDaMaster committed:
Update tgui/packages/tgfont/README_ICON_TUTORIAL.txt

Co-authored-by: Mothblocks <35135081+Mothblocks@users.noreply.github.com>

KirbyDaMaster reviewed (commented):
Outdated

Time-Green commented on code/datums/quirks/negative.dm:
```diff
@@ -447,13 +448,30 @@
+	var/mob/living/carbon/human/human_holder = quirk_holder
+	human_holder.del_and_replace_bodypart(BODY_ZONE_L_ARM, new/obj/item/bodypart/l_arm/robot/surplus(quirk_holder))
+	human_holder.del_and_replace_bodypart(BODY_ZONE_R_ARM, new/obj/item/bodypart/r_arm/robot/surplus(quirk_holder))
+	human_holder.del_and_replace_bodypart(BODY_ZONE_L_LEG, new/obj/item/bodypart/l_leg/robot/surplus(quirk_holder))
```
Params are (/obj/blabla/bodypart, special), this wont work

Time-Green commented on code/datums/quirks/negative.dm:
```diff
@@ -447,13 +448,30 @@
-	prosthetic.replace_limb(human_holder)
-	qdel(old_part)
-	human_holder.regenerate_icons()
+	human_holder.del_and_replace_bodypart(limb_slot, prosthetic)
```
Just put this in the switches tbh

Time-Green commented:
Posthumous verified

Mothblocks commented:
the verified is only minor changes that doesnt count

Time-Green commented:
> the verified is only minor changes that doesnt count

verified is more like a state of mind
